src/mpid/ch4/src/ch4_part.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpidimpl.h"
+
+static MPIR_Request *part_req_create(void *buf, int partitions, MPI_Count count,
+                                     MPI_Datatype datatype, int rank, int tag,
+                                     MPIR_Comm * comm, MPIR_Request_kind_t kind)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_Request *req = NULL;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PART_REQ_CREATE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PART_REQ_CREATE);
+
+    /* Set refcnt=1 for user-defined partitioned pattern; decrease at request_free. */
+    req = MPIR_Request_create_from_pool(kind, 0);
+    MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+
+    MPIR_Comm_add_ref(comm);
+    req->comm = comm;
+
+    MPIR_Datatype_add_ref_if_not_builtin(datatype);
+    MPIDI_PART_REQUEST(req, datatype) = datatype;
+
+    MPIDI_PART_REQUEST(req, rank) = rank;
+    MPIDI_PART_REQUEST(req, tag) = tag;
+    MPIDI_PART_REQUEST(req, context_id) = comm->context_id;
+    MPIDI_PART_REQUEST(req, buffer) = buf;
+    MPIDI_PART_REQUEST(req, count) = count;
+    MPIDI_PART_REQUEST(req, active_flag) = 0;
+    req->u.part.partitions = partitions;
+
+    /* Inactive partitioned comm request can be freed by request_free.
+     * Completion cntr increases when request becomes active at start. */
+    MPIR_cc_set(req->cc_ptr, 0);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPID_PART_REQ_CREATE);
+    return req;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPID_Psend_init(void *buf, int partitions, MPI_Count count,
+                    MPI_Datatype datatype, int dest, int tag,
+                    MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PSEND_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PSEND_INIT);
+
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, dest);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+    MPIDI_workq_vci_progress_unsafe();
+
+    *request = part_req_create((void *) buf, partitions, count, datatype, dest, tag, comm,
+                               MPIR_REQUEST_KIND__PART_SEND);
```

Comment:
Why not let the device create the request as pt2pt and RMA paths do?

Reply:
I guess you are suggesting that we let the `MPIDIG` or a native path create the request?

The reason now I create it at the device layer is that, I think they neither netmod nor shmmod nor am dependent. 

Maybe we can do something similar as that of `win_create`. See pseudo code below:
```
#ifdef MPIDI_CH4_DIRECT_NETMOD
    mpi_errno = MPIDI_NM_mpi_psend_init(...);
    MPIR_ERR_CHECK(mpi_errno);
#else
    mpi_errno = MPIDIG_mpi_psend_init(...);
    MPIR_ERR_CHECK(mpi_errno);
#endif
```

In `MPIDIG_mpi_psend_init`:
 (1) creates the request
 (2) set device-layer attributes
 (3) initialize AM resource/protocal
 (4) call `NM|SHM_mpi_psend_init_hook` for any nm|shm specific initialization 

What do you think?

Reply:
Whiling doing the refactoring, I feel that we still have to create a new pattern for partitioned pt2pt initialization, rather than completely following `win_create`. I.e., it is pt2pt, thus it might bring more benefit than windows if we allow SHM do optimization and skip generic AM. Thus, I'd keep the current `SHM|NM_psend_init` selection in `MPID_Psend_init`, but only move the request creation part to MPIDIG.

Besides, it is unsure how NM|SHM may use the hooks to do additional initialization. Thus, I'd skip the hook definition for now.

Reply:
I agree. 

Also now I realize the request we created in `psend_init` is not the same as request we create from the device -- it is more like a window. So in that sense, I think it make more sense to have the request creation at higher layer, potentially even at the mpir layer. At which layer do we allocate window?

I agree that we should skip the hook definition now. In general, I think we should delay any design until we need it in a implementation.

Reply:
Window is created at device layer. See `MPIDIG_mpi_win_create`->`win_init`

Reply:
I see. That feels a bit odd. I guess we designed that window synchronization only to be implemented in generic layer, so maybe that makes sense to have the window creation there. But frankly, I think that should be done at MPIR-layer, and call device hooks to populate fields if necessary.

With partitioned communication, do we plan to always have matching done via active messages? If yes, then I think we can follow the same pattern as windows since you can't skip AM anyway.

Reply:
Can't do window creation at MPIR because the RMA code in ch4 was completely different from that of ch3. Will be messy if we abstract some common parts of creation into MPIR but leave the remaining in device layer.

For partitioned pt2pt, we will need AM-based sync for both the potential RDMA version and the pt2pt version. But for each version, we may need exchange different attributes (e.g., `seq_no` for pt2pt version) and thus we will need different AM packet and request attributes.

Based on this thought, the attributes in `MPIDI_part_request_t` are generic for all approaches; the ones in `MPIDIG_part_request_t am` is AM-version specific. When a different version is added in future, the structure in `MPIDI_part_request_t` will look like
```
typedef struct MPIDI_part_request {
    union {
         MPIDIG_part_request_t am;
         xxx;  <- new version
    } u;

    /* partitioned attributes */
    void *buffer;
    int partitions;
    MPI_Aint count;             /* count per partition */
    int rank;
    int tag;
    MPIR_Context_id_t context_id;
    MPI_Datatype datatype;
} MPIDI_part_request_t;
```